src/core_plugins/elasticsearch/lib/check_es_version.js

```diff
@@ -1,31 +1,77 @@
+/**
+ * ES and Kibana versions are locked, so Kibana should require that ES has the same version as
+ * that defined in Kibana's package.json.
+ */
+
 import _ from 'lodash';
 import esBool from './es_bool';
-import versionSatisfies from './version_satisfies';
+import semver from 'semver';
+import isEsCompatibleWithKibana from './is_es_compatible_with_kibana';
 import SetupError from './setup_error';
 
-module.exports = function (server) {
+module.exports = function checkEsVersion(server, kibanaVersion) {
   server.log(['plugin', 'debug'], 'Checking Elasticsearch version');
 
   const client = server.plugins.elasticsearch.client;
-  const engineVersion = server.config().get('elasticsearch.engineVersion');
 
   return client.nodes.info()
   .then(function (info) {
-    const badNodes = _.filter(info.nodes, function (node) {
-      // remove nodes that satify required engine version
-      return !versionSatisfies(node.version, engineVersion);
-    });
+    // Aggregate incompatible ES nodes.
+    const incompatibleNodes = [];
 
-    if (!badNodes.length) return true;
+    // Aggregate ES nodes which should prompt a Kibana upgrade.
+    const warningNodes = [];
 
-    const badNodeNames = badNodes.map(function (node) {
-      return 'Elasticsearch v' + node.version + ' @ ' + node.http_address + ' (' + node.ip + ')';
+    _.forEach(info.nodes, esNode => {
```

Comment:
Rather than imperatively building these arrays in the same loop, how about just filtering?

```js
const incompatibleNodes  = info.nodes.filter(esNode => {
  return !isEsCompatibleWithKibana(esNode.version, kibanaVersion);
})

const warningNodes = info.nodes.filter(esNode => {
  return semver.gt(esNode.version, kibanaVersion);
});
```